### PR TITLE
Add grpc tooling to the project

### PR DESCRIPTION
**Pull request by r-marques**
- black excludes auto generated grpc files
- isort excludes auto generated grpc files
- pylint excludes auto generated grpc files
- added mypy stubs
- create tests
- added grpc helloworld example
- added custom command to setup.py to compile protofiles

---

protofiles are compiled when installing the project in development mode:
```
$ pip install -e .[dev]
```
Auto generated grpc files are committed to the repo. This way whoever is using this project does not need to have all the tooling to compile the proto files.

To test the gprc service:
```
$ python -m xain.grpc.numproto_server
NumProto server received: [0 1 2 3 4 5 6 7 8 9]
NumProto server sent: [ 0  2  4  6  8 10 12 14 16 18]

# another terminal
$ python -m xain.grpc.numproto_client
NumProto client sent: [0 1 2 3 4 5 6 7 8 9]
NumProto client received: [ 0  2  4  6  8 10 12 14 16 18]
```

In the example service the server receives a numpy array, doubles all the elements and returns the updated numpy array to the client.

---

There is a problem with using the the `cpp` implementation that results in the following error:
```
Traceback (most recent call last):
  File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/rodmar/code/autofl/xain/grpc/numproto_server.py", line 23, in <module>
    from xain.grpc import hellonumproto_pb2, hellonumproto_pb2_grpc
  File "/home/rodmar/code/autofl/xain/grpc/hellonumproto_pb2.py", line 26, in <module>
    dependencies=[numproto_dot_protobuf_dot_ndarray__pb2.DESCRIPTOR,])
  File "/home/rodmar/.virtualenvs/autofl/lib/python3.6/site-packages/google/protobuf/descriptor.py", line 879, in __new__
    return _message.default_pool.AddSerializedFile(serialized_pb)
TypeError: Couldn't build proto file into descriptor pool!
Invalid proto descriptor for file "xain/grpc/hellonumproto.proto":
  numproto/protobuf/ndarray.proto: Import "numproto/protobuf/ndarray.proto" has not been loaded.
  hellonumproto.NumProtoRequest.arr: "numproto.protobuf.NDArray" seems to be defined in "ndarray.proto", which is not imported by "xain/grpc/hellonumproto.proto".  To use it here, please add the necessary import.
  hellonumproto.NumProtoReply.arr: "numproto.protobuf.NDArray" seems to be defined in "ndarray.proto", which is not imported by "xain/grpc/hellonumproto.proto".  To use it here, please add the necessary import.
```

There are a lot of issues on github about this problem and no clear solution. For now I decided to just set the default implementation of protobuf to python.

I will investigate this issue further in another pr and try to switch back to the cpp implementation

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -17,7 +50,9 @@
     "botocore==1.12.220",  # Apache License 2.0
     "boto3==1.9.220",  # Apache License 2.0
     "awscli==1.16.230",  # Apache License 2.0
-    "faker==2.0.0",  # MIT License
```
Comment (r-marques) [inline setup.py]: need to fix `faker` dependency. it seems to have been moved from`dev_require` to `install_requires` in master

File: xain/grpc/numproto_server.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Copyright 2015 gRPC authors.
```
Comment (r-marques) [inline xain/grpc/numproto_server.py]: remove google disclaimers

File: protobuf/xain/grpc/hellonumproto.proto [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+service NumProtoServer {
+  rpc SayHelloNumProto (NumProtoRequest) returns (NumProtoReply) {}
+}
+
+message NumProtoRequest {
+  numproto.protobuf.NDArray arr = 1;
+}
+
+// The response message containing the greetings
```
Comment (r-marques) [inline protobuf/xain/grpc/hellonumproto.proto]: remove outdated comment

File: xain/grpc/test_grpc.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        NumProtoServer(), server
+    )
+    server.add_insecure_port("localhost:50051")
+    server.start()
+    yield
+    server.stop(0)
+
+
+# pylint: disable=W0613,W0621
+def test_greeter_server(greeter_server):
```
Comment (danieljanes) [inline xain/grpc/test_grpc.py]: Mark as integration test

File: setup.py [hunk 5]
```diff
@@ -1,11 +1,44 @@
...
+        command = [
+            "grpc_tools.protoc",
+            f"--proto_path={numproto_path}",
+            "--proto_path=./protobuf",
+            "--python_out=./",
+            "--grpc_python_out=./",
+            "--mypy_out=./",
+        ] + proto_files
+
+        print("Building proto_files {}".format(proto_files))
```
Comment (matklad) [inline setup.py]: As we require Python 3.6, we could use f-strings: `print(f"Building proto_files {proto_files}")`.

Comment (r-marques) [reply setup.py]: True. Currently we have both going on in the code.

At some point we should agree on a style guide.

**Pull request merged**